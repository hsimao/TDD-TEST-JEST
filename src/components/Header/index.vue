<template>
  <div class="header">
    <div class="header__content">
      TodoList
      <input @keyup.enter="handleInput"
        class="header__input"
        type="text"
        data-test="input"
        v-model.trim="inputValue"
        placeholder="TodoItem">
    </div>
  </div>
</template>

<script>
export default {
  name: 'Header',

  data() {
    return {
      inputValue: ''
    }
  },
  methods: {
    handleInput() {
      if (this.inputValue) {
        this.$emit('add', this.inputValue)
        this.inputValue = ''
      }
    }
  }
}
</script>

<style lang="scss" scoped>
.header {
  line-height: 60px;
  background-color: #333;

  &__content {
    width: 600px;
    display: flex;
    justify-content: space-between;
    align-items: center;
    margin: 0 auto;
    color: white;
    font-size: 24px;
  }

  &__input {
    width: 360px;
    outline: none;
    line-height: 24px;
    text-indent: 10px;
    color: #333;
  }
}
</style>
