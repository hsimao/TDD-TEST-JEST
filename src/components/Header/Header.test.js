import { shallowMount } from '@vue/test-utils'
import Header from './index.vue'
import { findTestWrapper } from '@/utils/testUtils'

it('Header 組件需要有 input 輸入欄位', () => {
  const wrapper = shallowMount(Header)
  const input = findTestWrapper(wrapper, 'input')

  expect(input.exists()).toBe(true)
})

it('input 初始值為空', () => {
  const wrapper = shallowMount(Header)
  const inputValue = wrapper.vm.$data.inputValue

  expect(inputValue).toBe('')
})

it('input 欄位發生變化時，inputValue 數據需要對應更新', () => {
  const wrapper = shallowMount(Header)
  const input = findTestWrapper(wrapper, 'input')
  input.setValue('Mars')
  const inputValue = wrapper.vm.$data.inputValue

  expect(inputValue).toBe('Mars')
})

it('input 欄位輸入 enter, 若為空值則不可以送出', () => {
  const wrapper = shallowMount(Header)
  const input = findTestWrapper(wrapper, 'input')

  // 模擬輸入空值，在觸發 enter 事件，來判斷是否有送出 $emit add 事件
  input.setValue('')
  input.trigger('keyup.enter')

  expect(wrapper.emitted().add).toBeFalsy() // toBeFalsy 沒有觸發
})

it('input 欄位輸入 enter, 若為有值時需要送出 add 事件, 同時清空 inputValue 值', () => {
  const wrapper = shallowMount(Header)
  const input = findTestWrapper(wrapper, 'input')

  // 模擬輸入空值，在觸發 enter 事件
  input.setValue('Mars')
  input.trigger('keyup.enter')

  expect(wrapper.emitted().add).toBeTruthy() // toBeTruthy 有觸發
  expect(wrapper.vm.$data.inputValue).toBe('')
})

// 樣式修改提示
it('Header DOM 節點快照發生改變, 請確認', () => {
  const wrapper = shallowMount(Header)
  expect(wrapper).toMatchSnapshot()
})
