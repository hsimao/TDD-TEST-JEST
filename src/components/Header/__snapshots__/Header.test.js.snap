// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header DOM 節點快照發生改變, 請確認 1`] = `
<div class="header">
  <div class="header__content">
    TodoList
    <input type="text" data-test="input" placeholder="TodoItem" class="header__input"></div>
</div>
`;
